File: sdks/local-rp/java/src/main/java/community/catalyst/linkkeys/localrp/ClaimError.java
package community.catalyst.linkkeys.localrp;

/** A claim signature/revocation/expiry verification failure. Mirrors {@code liblinkkeys::claims::ClaimError}. */
public class ClaimError extends RuntimeException {
    public enum Kind {
        UNSIGNED,
        KEY_NOT_FOUND,
        KEY_REVOKED,
        KEY_EXPIRED,
        UNSUPPORTED_ALGORITHM,
        SIGNATURE_INVALID,
        DOMAIN_KEYS_UNAVAILABLE,
        DOMAIN_UNVERIFIED,
        REVOKED,
        BAD_EXPIRY,
        EXPIRED,
    }

    private final Kind kind;
    private final String detail;

    public ClaimError(Kind kind, String detail) {
        super(detail == null ? kind.toString() : kind + ": " + detail);
        this.kind = kind;
        this.detail = detail;
    }

    public Kind kind() {
        return kind;
    }

    public String detail() {
        return detail;
    }
}

File: sdks/local-rp/csharp/src/LinkKeys.LocalRp/ClaimError.cs
namespace LinkKeys.LocalRp;

/// <summary>A claim signature/revocation/expiry verification failure. Mirrors <c>liblinkkeys::claims::ClaimError</c>.</summary>
public class ClaimError : Exception
{
    public enum ErrorKind
    {
        Unsigned,
        KeyNotFound,
        KeyRevoked,
        KeyExpired,
        UnsupportedAlgorithm,
        SignatureInvalid,
        DomainKeysUnavailable,
        DomainUnverified,
        Revoked,
        BadExpiry,
        Expired,
    }

    public ErrorKind Kind { get; }

    public string? Detail { get; }

    public ClaimError(ErrorKind kind, string? detail) : base(detail is null ? kind.ToString() : $"{kind}: {detail}")
    {
        Kind = kind;
        Detail = detail;
    }
}
